### PR TITLE
Sort: Use cub::WarpMergeSort for small sorts (32 < n <= 128)

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96223

We currently use `bitonicSortKVInplace` for sorts of size `n <= 32`
but use `radixSortKVInplace` for `32 < n <= 4096`. Bitonic sort is
also unstable, which forces stable sorts fall back to which is up to
4x slower in this small regime.

This PR adds a new kernel `warpMergeSortKVInplace` using
`cub::WarpMergeSort` to implement sorts with `32 < n <= 128` and all
stable sorts with `n < 128`. This results in up to a 2x speedup for
unstable sorts and up to 15x for stable sorts, depending on the input
geometry.

This also doesn't increase the total number of kernels since we are
replacing radix-sorts of size 32 and 128.

### CODE REVIEW

**Comment by peterbell10** [comment]
This shows the speedup for all sort length from 1 to 128, both as the first and last dim of a 2d tensor and for various batch sizes. In all cases the `WarpMergeSort` shows significant speedups:
unstable_sort

File: aten/src/ATen/native/cuda/Sort.cu [hunk 1]
```diff
@@ -28,12 +29,21 @@ static int minimum_grid_for_occupancy(T kernel, int max_block_size) {
...
 
-// For very small sorts, use bitonicSortKVInPlace which performs
-// better because it can sort multiple arrays within the same block of
-// threads, improving occupancy.
-//
-// TODO: cub in CUDA 11.6 has a WarpMergeSort primitive that could
-// replace the bitonic sort here.
+template <typename T>
+constexpr bool has_nan() {
+  if constexpr (std::numeric_limits<T>::is_specialized) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Sort.cu:34]: nice! I hope it won't be reverted because of internal VS2017 builds

**Review by ngimel — APPROVED** [review]

**Comment by osalpekar** [comment]
Seeing build failures on a large number of internal builds:
```
[2023-03-24T12:32:58.193-07:00] fbcode/caffe2/aten/src/ATen/native/cuda/SortUtils.cuh(224): error: union "<unnamed>" has no member "load_values"
[2023-03-24T12:32:58.193-07:00]           detected during:
[2023-03-24T12:32:58.193-07:00]             instantiation of "void at::native::warpMergeSortKVInPlace<KeyDims,ValueDims,sort_size,max_block_dim_y,K,V,Comparator,IndexType>(at::cuda::detail::TensorInfo<K, IndexType>, IndexType, IndexType, IndexType, at::cuda::detail::TensorInfo<V, IndexType>, IndexType, Comparator, K) [with KeyDims=-1, ValueDims=-1, sort_size=128, max_block_dim_y=16, K=c10::impl::ScalarTypeToCPPTypeT<c10::ScalarType::Char>, V=int64_t, Comparator=at::native::LTOp<c10::impl::ScalarTypeToCPPTypeT<c10::ScalarType::Char>, true>, IndexType=uint64_t]" 
[2023-03-24T12:32:58.193-07:00] fbcode/caffe2/aten/src/ATen/native/cuda/Sort.cu(132): here
[2023-03-24T12:32:58.193-07:00]             instantiation of "void at::native::WarpMergeSort<sort_size>::sort<A,K,V,IndexType>(at::cuda::detail::TensorInfo<K, IndexType>, IndexType, IndexType, IndexType, at::cuda::detail::TensorInfo<V, IndexType>, IndexType, __nv_bool) [with sort_size=128, A=-1, K=c10::impl::ScalarTypeToCPPTypeT<c10::ScalarType::Char>, V=int64_t, IndexType=uint64_t]" 
[2023-03-24T12:32:58.193-07:00] fbcode/caffe2/aten/src/ATen/native/cuda/Sort.cu(298): here
[2023-03-24T12:32:58.193-07:00]             instantiation of "void at::native::sortCommon(Sorter, const at::TensorBase &, const at::TensorBase &, int, __nv_bool) [with Sorter=at::native::WarpMergeSort<128>]" 
[2023-03-24T12:32:58.193-07:00] fbcode/caffe2/aten/src/ATen/native/cuda/Sort.cu(362): here
[2023-03-24T12:32:58.193-07:00]
```

More details: [D44378546](https://www.internalfb.com/diff/D44378546)

**Comment by peterbell10** [comment]
Is there more context in the build log? That error makes no sense unless there was an earlier failure.

**Comment by ngimel** [comment]
```
fbcode/caffe2/aten/src/ATen/native/cuda/SortUtils.cuh(201): error: namespace "at_cuda_detail::cub" has no member "WarpLoad"
[2023-03-24T12:15:49.505-07:00]           detected during:
[2023-03-24T12:15:49.505-07:00]             instantiation of "void at::native::WarpMergeSort<sort_size>::sort<A,K,V,IndexType>(at::cuda::detail::TensorInfo<K, IndexType>, IndexType, IndexType, IndexType, at::cuda::detail::TensorInfo<V, IndexType>, IndexType, __nv_bool) [with sort_size=128, A=-2, K=c10::impl::ScalarTypeToCPPTypeT<c10::ScalarType::Byte>, V=int64_t, IndexType=unsigned int]" 
[2023-03-24T12:15:49.505-07:00] fbcode/caffe2/aten/src/ATen/native/cuda/Sort.cu(298): here
[2023-03-24T12:15:49.505-07:00]             instantiation of "void at::native::sortCommon(Sorter, const at::TensorBase &, const at::TensorBase &, int, __nv_bool) [with Sorter=at::native::WarpMergeSort<128>]" 
[2023-03-24T12:15:49.505-07:00] fbcode/caffe2/aten/src/ATen/native/cuda/Sort.cu(362): here
[2023-03-24T12:15:49.505-07:00]
```
apparently earlier cuda versions (we have 11.4) don't have `WarpLoad`

**Comment by peterbell10** [comment]
@ngimel this change requires CUDA 11.6 but AFAIK the public release only supports CUDA 11.7+. Are there plans to update meta-internal users to newer CUDA versions, or do I need to hide this change behind some CUDA version checks?

**Comment by ngimel** [comment]
There are plans to update meta internal version, but there's no ETA, so if you could hide it behind ifdefs that would be great.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**